src/model_signing/_signing/sign_pkcs11.py

```diff
@@ -0,0 +1,304 @@
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections.abc import Iterable
+import pathlib
+from typing import Optional
+
+from asn1crypto.algos import DSASignature
+from asn1crypto.core import OctetString
+from asn1crypto.keys import ECDomainParameters
+from asn1crypto.keys import PublicKeyInfo
+from cryptography import x509
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.primitives.asymmetric import ec
+from google.protobuf import json_format
+import PyKCS11
+from sigstore_protobuf_specs.dev.sigstore.bundle import v1 as bundle_pb
+from sigstore_protobuf_specs.dev.sigstore.common import v1 as common_pb
+from sigstore_protobuf_specs.io import intoto as intoto_pb
+from typing_extensions import override
+
+from model_signing._signing import sign_sigstore_pb as sigstore_pb
+from model_signing._signing import signing
+from model_signing._signing.pkcs11uri import Pkcs11URI
+
+
+MODULE_PATHS: Iterable[str] = [
+    "/usr/lib64/pkcs11/",  # Fedora, RHEL, openSUSE
+    "/usr/lib/pkcs11/",  # Fedora 32 bit, ArchLinux
+]
+
+
+def _check_supported_ec_key(public_key: ec.EllipticCurvePublicKey):
+    """Checks if the elliptic curve key is supported by our package.
+
+    We only support a family of curves, trying to match those specified by
+    Sigstore's protobuf specs.
+    See https://github.com/sigstore/model-transparency/issues/385.
+
+    Args:
+        public_key: The public key to check. Can be obtained from a private key.
+
+    Raises:
+        ValueError: The key is not supported.
+    """
+    curve = public_key.curve.name
+    if curve not in ["secp256r1", "secp384r1", "secp521r1"]:
+        raise ValueError(f"Unsupported key for curve '{curve}'")
+
+
+def get_ec_key_hash(
+    public_key: ec.EllipticCurvePublicKey,
+) -> hashes.HashAlgorithm:
+    """Returns the public key hashing algorithm.
+
+    We need to record this in the sigstore bundle when signing and retrieve when
+    performing verification. This is according to sigstore protobuf specs and is
+    used both when signing with keys and when signing with certificates.
+
+    Args:
+        public_key: The public key to get the hash algorithm from.
+
+    Raises:
+        ValueError: The key is not supported.
+    """
+    key_size = public_key.curve.key_size
+
+    # TODO: Once Python 3.9 support is deprecated revert to using `match`
+    if key_size == 256:
+        return hashes.SHA256()
+    if key_size == 384:
+        return hashes.SHA384()
+    if key_size == 521:
+        return hashes.SHA512()
+
+    raise ValueError(f"Unexpected key size {key_size}")
```

Comment:
Can we reuse the ones from src/model_signing/_signing/sign_ec_key.py?

Like https://github.com/sigstore/model-transparency/blob/75cc528266278a28d8fe3d4df937244e8ed80035/src/model_signing/_signing/sign_certificate.py#L161

Reply:
Done.